pkg/hubble/relay/pool/client.go

```diff
@@ -55,11 +57,56 @@ func (b GRPCClientConnBuilder) ClientConn(target, hostname string) (poolTypes.Cl
 	} else {
 		// NOTE: gosec is unable to resolve the constant and warns about "TLS
 		// MinVersion too low".
-		tlsConfig := b.TLSConfig.ClientConfig(&tls.Config{ //nolint:gosec
+		baseConf := &tls.Config{ //nolint:gosec
 			ServerName: hostname,
 			MinVersion: hubbleopts.MinTLSVersion,
-		})
-		opts = append(opts, grpc.WithTransportCredentials(credentials.NewTLS(tlsConfig)))
+		}
+		opts = append(opts, grpc.WithTransportCredentials(
+			&grpcTLSCredentialsWrapper{
+				TransportCredentials: credentials.NewTLS(b.TLSConfig.ClientConfig(baseConf)),
+				baseConf:             baseConf,
+				TLSConfig:            b.TLSConfig,
+			},
+		))
 	}
 	return grpc.DialContext(ctx, target, opts...)
 }
+
+var _ credentials.TransportCredentials = &grpcTLSCredentialsWrapper{}
+
+// grpcTLSCredentialsWrapper wraps gRPC TransportCredentials and fetches the
+// newest TLS configuration from certloader whenever we a new TLS connection
+// is established.
+//
+// A gRPC ClientConn will call ClientHandshake whenever it tries to establish
+// a new TLS connection. This happens in the beginning when dialing, but also
+// when the connection is lost and gRPC tries to reestablish the connection.
+// Wrapping the ClientHandshake and fetching the updated certificate and CA,
+// allows us to transparently reload certificates when they change, without
+// closing and redialing the gRPC ClientConn.
+type grpcTLSCredentialsWrapper struct {
```

Comment:
Can you think of other cases where this might be needed? Might be useful to extract it into a helper package at some point. Hubble CLI comes to mind, but I'm not entirely sure.

Reply:
True that might be useful for the Hubble CLI, although I don't think it does cert reloading anyways. And I don't know of any other gRPC client we have right now that would need certificate reloading.

Reply:
Ok, I think this is fine then.